components/widgets/menu/menu_items/submenu_item_action.jsx

```diff
@@ -0,0 +1,140 @@
+        icon: PropTypes.node,
+        action: PropTypes.func.isRequired,
+        openModal: PropTypes.func,
+        xOffset: PropTypes.number,
+        ariaLabel: PropTypes.string,
+        root: PropTypes.bool,
+    };
+    static defaultProps = {
+        show: true,
+    };
+
+    constructor(props) {
+        super(props);
+        this.node = React.createRef();
+
+        this.state = {
+            show: false,
+        };
+    }
+
+    show = () => {
+        this.setState({show: true});
+    }
+
+    hide = () => {
+        this.setState({show: false});
+    }
+
+    handleShowMobileSubMenuItem = (elements, action) => {
+        const openModal = this.props.openModal;
+        const submenuModalData = {
+            ModalId: ModalIdentifiers.PLUGIN_MOBILE_SUBMENU,
+            dialogType: SubMenuModal,
+            dialogProps: {
+                elements,
+                action,
+                openModal,
+            },
+        };
+
+        this.props.openModal(submenuModalData);
+    }
+
+    render() {
+        const {id, text, subMenu, icon, action, xOffset, ariaLabel, root} = this.props;
+        const isMobile = Utils.isMobile();
+
+        let textProp = text;
+        if (icon) {
+            textProp = (
+                <React.Fragment>
+                    <span className='icon'>{icon}</span>
+                    {text}
+                </React.Fragment>
+            );
+        }
+
+        const parentWidth = this.node && this.node.current ? this.node.current.getBoundingClientRect().width - 19 : 0;
+
+        const subMenuStyle = {
+            visibility: this.state.show && subMenu && subMenu.length && !isMobile ? 'visible' : 'hidden',
+            right: (parseInt(xOffset, 10) - 1) + 'px',
+        };
+
+        return (
+            <li
+                className='MenuItem'
+                role='menuitem'
+                id={id + '_menuitem'}
+                ref={this.node}
+            >
+                <div
+                    id={id}
+                    aria-label={ariaLabel}
+                    onMouseEnter={this.show}
+                    onMouseLeave={this.hide}
+                    onClick={(event) => {
+                        if (isMobile) {
+                            if (subMenu && subMenu.length) {
+                                if (!root) {
+                                    event.stopPropagation();
+                                }
+                                this.handleShowMobileSubMenuItem(subMenu, action);
+                            } else {
+                                action(id);
+                            }
+                        } else if (event.nativeEvent.path &&
+                                    event.nativeEvent.path.slice(0, 2).find((e) => e.id === id)) {
+                            action(id);
+                        } else if (!event.nativeEvent.path) {
+                            action(id);
+                        }
+                    }}
```

Comment:
I think it's worth moving this `onClick` handler to it's own function on the top with some comments on the special cases so it's more readable of what it is doing here (maybe even trying to combine some conditions at the end given that they all execute `action(id)`

Reply:
Done.  Added comments as well, but didn't collapse the logic as it is as clear as I can currently make it.  Open to suggestions to improving the functionality or readability.